### PR TITLE
Core Architecture Stability & Performance Overhaul

### PR DESCRIPTION
**Pull request by Cheralia**
This PR addresses severe training instabilities, runaway energy metrics, and massive data redundancies in the Predictive Coding (PC) network. By introducing exact probability-distribution math at the output layer, normalizing gradient updates, fixing multi-head attention weight slicing, and streamlining the inference loop, this update drastically accelerates epoch times and allows the model's energy and perplexity to converge smoothly.

### 1. Data Pipeline & Loss Function Enhancements

* **Dataloader Optimization (Eliminated Extreme Redundancy):** Updated the sequence extraction stride from `stride=1` to match the `block_size` (e.g., 208). Previously, a stride of 1 shifted tokens by a single position, creating 99% overlapping context windows and ballooning batches to over 30,000 per epoch. The data is now chopped into continuous chunks with exactly 1 token of overlap, drastically reducing batches to ~145 without sacrificing dataset coverage, resulting in vastly faster epoch times.
* **Switched Output Layer to KL-Divergence:** Changed `output_energy_fn_name` from Mean Squared Error (`pc_e`) to KL-Divergence (`kld`). Since language modeling is a discrete classification task, KLD properly measures statistical distance at the output layer, providing a mathematically sound learning signal.
* **Unclamped Energy & Simplified Gradients:** Removed artificial `torch.clamp` caps on KLD energy calculation, allowing the true error magnitude to flow backward and preventing the optimizer from stalling. Replaced the heavy Softmax vector-Jacobian product with its direct, mathematically exact derivative (`target - probs`), which makes the backward pass faster and more numerically stable.

### 2. Gradient Normalization & Learning Phase Fixes

* **Fixed the "Sum vs. Mean" Gradient Explosion:** The manual weight update using `torch.einsum` was summing gradients across the Batch and Sequence dimensions instead of averaging them, causing massive weight overshoots and infinitely exploding energy. Divided the calculated update by `(Batch * Sequence)` to properly normalize the `local_lr` step size.
* **Resolved Destructive Learning:** Previously, the network updated weights at every time step of the inference loop ($t=0$ to $t=T-1$), meaning weights were updated based on unconverged, noisy error signals. Weight updates are now restricted to the final iteration of the loop, ensuring the latent states have settled into an optimal representation before updating the network parameters.

### 3. Attention Mechanism Fixes

* **Corrected Attention Dimension Slicing:** In `step_attn`, top-down errors routing back to multi-head attention weights were being sliced on the wrong axis (`dim=1` instead of `dim=0`). Since `nn.Linear` stores weights as `[out_features, in_features]`, the previous code scrambled updates across heads. This fix ensures errors from specific attention heads correctly update their respective weights.

### 4. Embedding & Initialization Enhancements

* **Removed LayerNorm from Embedding Layer:** Dropped LayerNorm from the raw embeddings. Because we were skipping the LayerNorm Jacobian during the manual backward pass, the raw embedding weights were inflating infinitely (as LayerNorm continuously scaled them down while additive error kept pushing them up). Furthermore, the target state from the attention layer above is pre-norm, so removing LayerNorm ensures apples-to-apples state comparisons.
* **Introduced Sequential Bottom-Up Warmup:** Instead of initializing latent states (`x`) to pure random noise, a quick sequential PyTorch forward pass is now run before the parallel $T$ loop. This populates `_x_cache` with actual network predictions, giving the top-down predictive coding inference a grounded starting point to fine-tune from, rather than requiring several time steps just to propagate signal.

## Impact

* **Speed:** Epochs train exponentially faster due to dataloader stride optimization and direct Jacobian simplifications.
* **Stability:** Eliminates exploding energy and catastrophic perplexity spikes during training by enforcing proper $T$-step convergence before weight updates and normalizing the `einsum` math.
* **Correctness:** Ensures multi-head attention properly isolates weight updates per head and the output layer properly models discrete probabilities.

### CODE REVIEW

File: utils/config_utils.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def load_best_config():
         "batch_size": 8,
         "num_epochs": 5,
         "internal_energy_fn_name": "pc_e",
-        "output_energy_fn_name": "pc_e",
+        "output_energy_fn_name": "kld",
```
Comment (quinbez) [inline utils/config_utils.py:35]: `output_energy_fn_name` should be changed to "kld" on tuning too. Otherwise the tuning runs will still use `pc_e` for the output layer, which is inconsistent with the change made in the fallback config.

Comment (Cheralia) [reply utils/config_utils.py:35]: Yes, good catch. Done.
